sql_metadata/parser.py

```diff
@@ -0,0 +1,533 @@
+from sqlparse.tokens import Name, Number, Punctuation, Whitespace, Wildcard
+
+from sql_metadata.generalizator import Generalizator
+from sql_metadata.keywords_lists import (
+    FUNCTIONS_IGNORED,
+    KEYWORDS_BEFORE_COLUMNS,
+    KEYWORDS_IGNORED,
+    SUBQUERY_PRECEDING_KEYWORDS,
+    TABLE_ADJUSTMENT_KEYWORDS,
+    WITH_ENDING_KEYWORDS,
+)
+from sql_metadata.token import EmptyToken, SQLToken
+from sql_metadata.utils import UniqueList
+
+
+class Parser:  # pylint: disable=R0902
+    """
+    Main class to parse sql query
+    """
+
+    def __init__(self, sql: Optional[str] = None) -> None:
+        self._raw_query = sql
+        self._query = self._preprocess_query()
+
+        self._tokens = None
+
+        self._columns = None
+        self._columns_dict = None
+
+        self._tables = None
+        self._table_aliases = None
+
+        self._with_names = None
+        self._subqueries = None
+        self._subqueries_names = None
+
+        self._limit_and_offset = None
+
+        self._values = None
+        self._values_dict = None
+
+    @property
+    def query(self) -> str:
+        """
+        Returns preprocessed query
+        """
+        return self._query
+
+    @query.setter
+    def query(self, value):
+        """
+        Resets cached attributes before setting new query
+        """
+        self._tokens = None
+
+        self._columns = None
+        self._columns_dict = None
+
+        self._tables = None
+        self._table_aliases = None
+
+        self._with_names = None
+        self._subqueries = None
+        self._subqueries_names = None
+
+        self._limit_and_offset = None
+
+        self._values = None
+        self._values_dict = None
+
+        self._raw_query = value
+        self._query = self._preprocess_query()
+
+    @property
+    def tokens(self) -> List[SQLToken]:
+        """
+        :rtype: list[SQLToken]
+        """
+        if self._tokens is not None:
+            return self._tokens
+
+        parsed = sqlparse.parse(self.query)
+        tokens = []
+        # handle empty queries (#12)
+        if not parsed:
+            return tokens
+
+        sqlparse_tokens = TokenList(parsed[0].tokens).flatten()
+        non_empty_tokens = [
+            token for token in sqlparse_tokens if token.ttype is not Whitespace
+        ]
+        last_keyword = None
+        subquery_level = 0
+        open_parenthesises = []
+        for index, tok in enumerate(non_empty_tokens):
+            token = SQLToken(
+                value=tok.value,
+                is_keyword=tok.is_keyword,
+                is_name=tok.ttype is Name,
+                is_punctuation=tok.ttype is Punctuation,
+                is_dot=str(tok) == ".",
+                is_wildcard=tok.ttype is Wildcard,
+                is_integer=tok.ttype is Number.Integer,
+                is_float=tok.ttype is Number.Float,
+                is_left_parenthesis=str(tok) == "(",
+                is_right_parenthesis=str(tok) == ")",
```

Comment:
I'm wondering if we shouldn't rather pass `tok` into the constructor of SQLToken to move this logic into the token module. Parser is already quite complex 🙂

Reply:
Yeah sounds like a good idea :)

Reply:
Maybe I will wait for your further review to not change anything major while you are still in the process. ;)

Reply:
Sure, sure. Let's tackle this one is another PR. And limit the scope of this one. It's already quite big, isn't? 😉

Reply:
Yeah, sorry for that :D it was like some changes caused new bugs or I found a functionality that  and the scope exploded a little 😅